src/core/lib/promise/for_each.h

```diff
@@ -77,8 +77,8 @@ struct NextValueTraits<T, absl::void_t<typename T::value_type>> {
     return NextValueType::kEndOfStream;
   }
 
-  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION static Value& MutableValue(T& t) {
-    return *t;
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION static Value&& TakeValue(T& t) {
```

Comment:
No change in `ForEach`, but the `Take` instead of a `move(Mutable())` permits construction of the taken value (and that allows some optimizaitons for NextMessage)